rfcs/0143-nix-store-acls.md

```diff
@@ -0,0 +1,195 @@
+
+For paths external to the Nix sandbox (added via `nix store add-{file,path}`, paths in the Nix language, `builtins.fetch*`, or flake sources), we add the user to the list when they send the path to the daemon.
+We might need to add a flag (like `--protect`) for the selective ACL mode.
+
+For derivations themselves (.drv files), we add the user to the list when they send the derivation to the daemon.
+
+For derivation outputs, we add user to the list when the user requests to realise the derivation and has access to the transitive dependencies, including the sources.
+
+When `selective-acl` is enabled, protected paths should not be readable by anyone during the build.
+Necessary permissions are granted after the build.
+
+There also should be a way to enable the protection for selective ACLs (perhaps `nix store access protect`), and explicitly grant (`nix store access grant`) or revoke (`nix store access revoke`) access of certain user or group to each individual path.
+Naturally, this should only be available to `trusted-user`s and users who have access to this path already (either because they are on the ACL, or they belong to a group on the ACL).
+
+### `db.sqlite`
+
+The `ValidPaths` table should get a new field, something like `allowed-users`, which should list the users and groups who have either provided proof of source for this derivation or had been explicitly granted access.
+When `selective-acl` is enabled, there should be a field like `protected` to mark a derivation as protected.
+
+### Changes to the system
+
+We should implement a way to restrict access to all the store paths for users.
+A first "line of defense" could be something like [RFC 97], which makes the store non-world-listable.
+However, it is separate from this RFC, and is not required.
+On top of that, we must enforce a stricter access control, using [POSIX ACLs](https://man7.org/linux/man-pages/man5/acl.5.html) to only allow users access to store paths if they are part of the ACL (or belong to a group on the ACL) for that path.
+Nix daemon (or other local store implementations) should execute the appropriate `setfacl` calls whenever a path is added to store or gets different permissions in the db.
+
+[RFC 97]: https://github.com/NixOS/rfcs/pull/97
+
+### Nix language
+
+We change the `derivation` and `path` builtins, such that a `__permissions` attribute, when set, is expected to be a list of strings, representing the users and groups who should be granted access to the store path (or the derivation outputs, in the case of `derivation`).
+
+If this list is set, the Nix daemon (or other store implementation) should be notified that the path/derivation is supposed to be protected (perhaps using a new worker protocol command), and after the build is completed, the ACLs should be set appropriately (only the users specified in the `__permissions` list should have access to the path).
```

Comment:
> Should that attribute have an impact on the derivation hash? It feels like it shouldn't (as it only touches the metadata), but at the same time not doing so wouldn't be properly backwards-compatible

It probably shouldn't. The lack of backwards-compatibility is annoying, but since `__permissions` wouldn't be `toString`-able, it would result in a clear failure on older Nix, which is a fine behaviour IMHO.

Reply:
> should bob be given access to "${./somePath}" since it's in the closure of that derivation output?

I don't think so; IMHO, Permissions are applied only to the object on which they are specified.

> And what if I don't add myself to the __permissions field?

Perhaps there should be a check to automatically add the evaluating user to the __permissions.